packages/utils/src/assets/asset-id.ts

```diff
@@ -0,0 +1,77 @@
+import { CryptoAsset, CryptoAssetId, Sip9Asset } from '@leather.io/models';
+
+import { assertUnreachable } from '../index';
+
+export function matchesAssetId(asset: CryptoAsset, assetId: CryptoAssetId) {
+  return getAssetId(asset).protocol === assetId.protocol && getAssetId(asset).id === assetId.id;
+}
+
+export function getAssetId(asset: CryptoAsset): CryptoAssetId {
+  switch (asset.protocol) {
+    case 'nativeBtc':
+      return {
+        protocol: 'nativeBtc',
+        id: asset.symbol,
+      };
+    case 'nativeStx':
+      return {
+        protocol: 'nativeStx',
+        id: asset.symbol,
+      };
+    case 'sip10':
+      return {
+        protocol: 'sip10',
+        id: asset.assetId,
+      };
+    case 'rune':
+      return {
+        protocol: 'rune',
+        id: asset.runeName,
+      };
+    case 'brc20':
+      return {
+        protocol: 'brc20',
+        id: asset.symbol,
+      };
+    case 'src20':
+      return {
+        protocol: 'src20',
+        id: asset.id,
+      };
+    case 'stx20':
+      return {
+        protocol: 'stx20',
+        id: asset.symbol,
+      };
+    case 'inscription':
+      return {
+        protocol: 'inscription',
+        id: asset.id,
+      };
+    case 'stamp':
+      return {
+        protocol: 'stamp',
+        id: asset.stamp.toString(),
+      };
+    case 'sip9':
+      return buildSip9AssetId(asset);
+    default:
+      assertUnreachable(asset);
+  }
+}
+
+export function buildSip9AssetId(asset: Sip9Asset): CryptoAssetId {
+  return {
+    protocol: asset.protocol,
+    id: `${asset.assetId}|${asset.tokenId}`,
+  };
+}
+
+export function readSip9IdFields(assetId: CryptoAssetId): {
```

Comment:
There would be a bit of an overhead with more types, but would be worth having `readSip9IdFields` require a narrower type instead of accepting a broader one and throwing an error? The name implies it expects Sip9, is there a case where an arbitrary assetId would be passed into this?